src/NewRemoting/MessageHandler.cs

```diff
@@ -41,10 +59,215 @@ public MessageHandler(InstanceManager instanceManager, FormatterFactory formatte
+									break;
+								case 3:
+									proxyType =
+										typeof(DelegateFuncProxyOnClient<,,>).MakeGenericType(arguments.ToArray());
+									break;
+								case 4:
+									proxyType =
+										typeof(DelegateFuncProxyOnClient<,,,>).MakeGenericType(arguments.ToArray());
+									break;
+								case 5:
+									proxyType =
+										typeof(DelegateFuncProxyOnClient<,,,,>).MakeGenericType(arguments.ToArray());
+									break;
+								default:
+									throw new NotSupportedException(
+										$"Unsupported number of arguments for function ({arguments.Count}");
+							}
+						}
+						else
+						{
+							hasReturnValue = false;
+							switch (arguments.Count)
+							{
+								case 0:
+									proxyType = typeof(DelegateProxyOnClient);
+									break;
+								case 1:
+									proxyType = typeof(DelegateProxyOnClient<>).MakeGenericType(arguments.ToArray());
+									break;
+								case 2:
+									proxyType = typeof(DelegateProxyOnClient<,>).MakeGenericType(arguments.ToArray());
+									break;
+								case 3:
+									proxyType = typeof(DelegateProxyOnClient<,,>).MakeGenericType(arguments.ToArray());
+									break;
+								case 4:
+									proxyType = typeof(DelegateProxyOnClient<,,,>).MakeGenericType(arguments.ToArray());
+									break;
+								default:
+									throw new NotSupportedException(
+										$"Unsupported number of arguments for action ({arguments.Count}");
+							}
+						}
+
+						var proxy = Activator.CreateInstance(proxyType);
+						var addEventMethod = proxyType.GetMethod("add_Event");
+						addEventMethod.Invoke(proxy, new[] { del });
+						_instanceManager.AddInstance(proxy, instanceId, referencesWillBeSentTo, proxyType);
+
+						w.Write((int)RemotingReferenceType.AddEvent);
+						LogMsg(RemotingReferenceType.AddEvent);
+						w.Write(instanceId);
+						w.Write(hasReturnValue);
+						// If the type of the delegate method is generic, we need to provide its type arguments
+						w.Write(arguments.Count);
+						foreach (var argType in arguments)
+						{
+							string arg = argType.AssemblyQualifiedName;
+							if (arg == null)
+							{
+								throw new RemotingException("Unresolved generic type or some other undefined case");
+							}
+
+							w.Write(arg);
+						}
+					}
+				}
+				else
+				{
+					// The delegate target is a static method
+					////						w.Write(string.Empty);
+					throw new InvalidOperationException("The delegate target is a static method");
+				}
+			}
+			else if (calledMethod.IsSpecialName && calledMethod.Name.StartsWith("remove_", StringComparison.Ordinal))
+			{
+				if (del.Target != null)
+				{
+					string instanceId = _instanceManager.GetDelegateTargetIdentifier(del, remoteInstanceId);
+
+					if (_instanceManager.TryGetObjectFromId(instanceId, out var existingDelegateProxy))
+					{
+						// Create a proxy class that provides a matching event for our delegate
```

Comment:
Comment is incorrect (we're not creating proxies here, we're attempting to delete them)